### PR TITLE
adding redux

### PR DESCRIPTION
**Pull request by hkasemir**
Re: https://github.com/stasm/alfred/issues/20
Ok, take a look - I've added some basic capability and restructuring for redux - still clearly lots of work to do :) Handing over to you to take if you wish, or if you'd prefer to leave comments, that's also good. 

basics: 
- I recommend putting all api requests in the `alfred-api-service.js` and turning these into async actions. All actions can be mapped to props through the `mapDispatchToProps` function, making the api requests run and set the redux state with their results.
- I took a shot at the basic state structure, but feel free to modify this as you see fit. I'm not fully clear on the way you're ideally setting this up.

I was struggling a little with getting the reports/updates created properly in the dev environment, so hopefully it works for you, as I was kind of guessing on a couple things.

### CODE REVIEW

**Comment by stasm** [comment]
Awesome, thanks @hkasemir!  I'll take a deeper look later today and leave comments here.

File: www/src/app/services/alfred-api-service.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        nextReportDate: new Date(nextReport.reportDate),
+        nextReportSlug: nextReport.slug,
+        updates: [...current, ...prev, ...next]
+      }
+    }).catch(console.error);
+}
+
+export function postReport() {
+  const date = new Date();
+  return post(`${API_URL}/reports`, {date});
```
Comment (stasm) [inline www/src/app/services/alfred-api-service.js]: What is the purpose of this API call?

File: www/src/app/services/alfred-api-service.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { API_URL } from '../../config';
+import { get, post } from '../utils';
+
+
+export function fetchUpdatesByAuthor(author) {
```
Comment (stasm) [inline www/src/app/services/alfred-api-service.js]: I think we'll also need a service function of getting updates for a given report. See `www/src/app/views/ReportContainer`.

File: www/src/app/actions/index.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import * as alf from '../services/alfred-api-service';
+
+export const types = {
+  SET_AUTHOR: 'SET_AUTHOR',
+  FETCH_UPDATES_BY_AUTHOR: 'FETCH_UPDATES_BY_AUTHOR',
+  FETCH_UPDATES_BY_AUTHOR_COMPLETED: 'FETCH_UPDATES_BY_AUTHOR_COMPLETED',
+  POST_REPORT: 'POST_REPORT',
+  POST_REPORT_COMPLETED: 'POST_REPORT_COMPLETED'
```
Comment (stasm) [inline www/src/app/actions/index.js]: What are these last two actions responsible for?

File: www/src/app/reducers/updates.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { types } from '../actions';
+
+const defaultState = {
+  inbox: [],
+  done: [],
+  goal: []
```
Comment (stasm) [inline www/src/app/reducers/updates.js]: if we choose to store updates in multiple arrays, one for each possible status, the code in `InboxContainer.handle*` methods will get a bit more complex.  For each change made by the user, we'll need to remove the update from the array corresponding to its previous status and push into the array corresponding to the new status.

Instead, can rather we keep all updates in a single `updates` array, independent of their status?  This makes it easier to change the status and I think it gives us more flexibility right now.

I'm not opposed to changing the layout of the data store, but perhaps it can be done in another PR.

File: www/src/app/services/alfred-api-service.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  const updatesByAuthor = `${API_URL}/updates?author=${author}`;
+  const currentReportsInfo = `${API_URL}/reports/current`;
+
+  return get(currentReportsInfo).then(([prevReport, nextReport]) => {
```
Comment (stasm) [inline www/src/app/services/alfred-api-service.js]: Should we move the action of getting the current reports info into its own service function, and possible its own Redux action which is dispatched when `InboxContainer` is mounted or when the inbox route is entered?

File: www/src/app/views/InboxContainer.js [hunk 6]
```diff
@@ -223,6 +219,22 @@ export default class InboxContainer extends Component {
...
       />
+      </section>
     );
   }
 }
+
+const mapStateToProps = (state) => {
+  return {
+    author: state.user.author,
+    updates: state.updates.inbox
```
Comment (stasm) [inline www/src/app/views/InboxContainer.js]: `inbox` is just one possible status an update can have.  The other are: `event`, `todo`, `done`, `goal`, `struggle`, `achievement`, `archive` and `removed`. I think this line should simply read:

```
updates: state.updates
```

File: www/src/app/views/InboxContainer.js [hunk 7]
```diff
@@ -172,43 +164,47 @@ export default class InboxContainer extends Component {
...
         )
       })
     );
   }
 
   render() {
+    const { postReport, updates, author } = this.props;
     return (
+      <section>
+      <button onClick={postReport}>post report</button>
```
Comment (stasm) [inline www/src/app/views/InboxContainer.js]: There's `postReport` here again and I'm not sure what it's used for.

File: www/src/app/views/InboxContainer.js [hunk 8]
```diff
@@ -1,18 +1,16 @@
...
+import { makeUpdate, post } from '../utils';
 
-export default class InboxContainer extends Component {
+class InboxContainer extends Component {
   constructor(props) {
     super(props);
 
-    const { author } = this.props.params;
-
     this.state = {
```
Comment (stasm) [inline www/src/app/views/InboxContainer.js]: Should we continue to use `this.state` and `this.setState` next to using Redux?  Do you mean this as a temporary measure to split the migration into smaller chunks?

**Review by stasm — COMMENTED** [review]
This is looking great, thanks!

I think the biggest questions I have right no revolve around using `this.setState` in addition to using Redux, as well as the layout of the store.

File: www/src/app/actions/index.js (hunk 3 above)
Comment (hkasemir) [reply www/src/app/actions/index.js]: This was an attempt to tie into the `postReport` function so that I could populate my inbbox with the reports necessary to not break the inbox... I'm not fluent with how the bot works, nor mongodb... so I tried this and left it in here, as it quasi-worked. This may be able to be refactored for something more useful in the future.

File: www/src/app/reducers/updates.js (hunk 4 above)
Comment (hkasemir) [reply www/src/app/reducers/updates.js]: I would argue with how redux works, that we can tie the actions into reducers to make moving from one array to another relatively trivial. That being said, I am happy to leave the store structure to you - this is your brainchild 😄 I would be happy to try to find time later this week to discuss the best way to set the store up. This PR was a first pass with the main goals of getting a few redux actions to dispatch, and integrate the dev-tools so we could easily visualize the store and have that to start the conversation :boom:

File: www/src/app/services/alfred-api-service.js (hunk 2 above)
Comment (hkasemir) [reply www/src/app/services/alfred-api-service.js]: yup yup! This was definitely a partial migration. I also think that most if not all of the `InboxContainer.handle*` methods would be better suited to redux actions. Doing it this way is likely to clean up the component code significantly.

File: www/src/app/services/alfred-api-service.js (hunk 5 above)
Comment (hkasemir) [reply www/src/app/services/alfred-api-service.js]: since there is so little documentation on the APIs, I'm not sure I would have the knowledge to quickly do this. This is part of why this PR has so many holes - I figured once I kicked off and got started with the redux conventions that you may be able to take over defining which calls would be best to do when.

File: www/src/app/services/alfred-api-service.js (hunk 1 above)
Comment (hkasemir) [reply www/src/app/services/alfred-api-service.js]: convenience during dev explained above - not proud of it 😹

File: www/src/app/views/InboxContainer.js (hunk 8 above)
Comment (hkasemir) [reply www/src/app/views/InboxContainer.js]: Again, I mostly left this here to be able to quickly do the partial migration, and start the conversation of what this will turn into. There are definitely cases where a component's state are useful. The way I see it is this:
- If it's info shared among different views, or a representation of what's in the DB, it should be set in the redux store and pulled into the view via `mapStateToProps`,
- If it's data/functionality for a simple component, it should be passed through explicitly via a prop on that component,
- if it's data of interest only to that component that is likely to change (edit text input values, for example) it should be a state variable of the component/its container.

File: www/src/app/views/InboxContainer.js (hunk 6 above)
Comment (hkasemir) [reply www/src/app/views/InboxContainer.js]: definitely interested in discussing your vision for the store model here! I think I don't fully grasp the functionality of different items, and how you want to structure them - let's see if we can arrange a time to chat about it.

**Review by hkasemir — COMMENTED** [review]
😄

File: www/src/app/services/alfred-api-service.js (hunk 1 above)
Comment (zbraniecki) [reply www/src/app/services/alfred-api-service.js]: I think it should not be a part of this PR and instead we may want to get a separate PR to add report management from the dashboard UI

**Comment by zbraniecki** [comment]
My notes:
- Let's move the postReport to a separate PR
- I'm with stas on the idea that item should have a status instead of being moved between collections. Unless Redux has any specific best practices with regards to this.

**Comment by stasm** [comment]
I think we can try having separate collections for each status, but that sounds like more work.  Let's change one thing at a time.

I'll create a document describing the data model I have in mind so that we can all discuss it.

**Comment by zbraniecki** [comment]
So, what's the next step here? I'd love this to land before it bitrots...

@hkasemir - can you remove the postReport pieces and remaining feedback from stas?

**Comment by stasm** [comment]
I'd like to test this before I merge but I don't think I'll have time today.  I'll try to get to it tomorrow.  If you can update the PR to my feedback by then that would be great! Thanks.

**Comment by hkasemir** [comment]
I would be happy to - though I would feel better having the data store model you were saying you might document @stasm so I can format the reducers to produce it. Let me know if you can come up with something here so I can work on it after work tonight!

**Comment by stasm** [comment]
I created a stub of the documentation here: https://github.com/stasm/alfred/wiki/Data-Store.  Let me know if you have questions about it.  The central piece is the `resolve` field which essentially serves as a naive version control system:  resolved items don't require any action and they're hidden from your inbox but they can still show up in previous reports.  The idea being that marking a goal from week A as an achievement in week B should still make the update show up as a goal in week A's report.

I plan to change `resolved` from a boolean to an integer: `0` means _unresolved_, `1` means _resolved and show it on reports_, `2` means _resolved and don't show it on reports_.  This is needed to make it possible to remove things from reports.  I mention this here just so that everyone is aware;  I'll implement this after your PR lands.

Thanks again, @hkasemir!

**Comment by hkasemir** [comment]
ok - I tried to structure the redux store exactly as you have it in the wiki - I hope it's closer to what you're looking for 😄 

major things I did to update this pr:
- removed the `postReport` stuff
- broke out the `fetchUpdatesByAuthor` stuff in alfred-api-service into 4 different functions so it's explicit what each call is doing. This makes it (in my opinion) much easier to understand and read. I also like this way of making each API call its own function so it's more extensible.
- I developed the actions and reducers to fit the new api-service
- I turned the Inbox component into a react class so I could connect redux to it.

Again, there is still a lot of work to be done to migrate this fully to redux, but I hope you can see the benefits of what we've managed so far!

Let me know if you have questions or see anything you think should be changed.

**Comment by hkasemir** [comment]
I'll be completely honest - this actually breaks a bit of functionality in the inbox since I took out the way you were managing state... looking into an intermediate fix for this now - as migrating the whole thing to redux to get it to work might take a bit longer.

**Comment by hkasemir** [comment]
ok, I was able to begin migrating UpdateList editables to using redux - including onStartEdit, onBlur, and onSubmit functions. I feel Like I kind of started off down a path that's relatively different from where the UpdateList component started, but I felt that dealing with all the different `handle*` functions would be simpler this way - i.e. using them to pull actions in via `connect` and `mapDispatchToProps`.

Eager to hear what you think! and yes, I'm sorry for breaking some of the functionality. 🙏

**Comment by stasm** [comment]
I'll review this later tonight.  Ideally, we'd find a way to start using Redux for a subset of actions and keep using `setState` for the rest, without compromising the functionality. We can then work on migrating the remaining actions in future PRs.

Thanks for all the hard work on this!

File: www/src/app/actions/index.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  // user
+  SET_AUTHOR:                        'SET_AUTHOR',
+
+  //ui
+  SET_EDITING:                       'SET_EDITING',
+  CANCEL_EDITING:                    'CANCEL_EDITING',
+
+  // updates
+  FETCH_UPDATES_BY_AUTHOR:           'FETCH_UPDATES_BY_AUTHOR',
+  FETCH_UPDATES_BY_AUTHOR_COMPLETED: 'FETCH_UPDATES_BY_AUTHOR_COMPLETED',
```
Comment (stasm) [inline www/src/app/actions/index.js]: This is purely subjective, but I like the convention of calling the async actions `REQUEST_FOO` and `RECEIVE_FOO`.  What do you think about it?

File: www/src/app/actions/index.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+export function setEditing(update) {
+  return createAction(types.SET_EDITING, update);
+}
+
+export function cancelEditing(update) {
+  return createAction(types.CANCEL_EDITING, update);
+}
+
+export function fetchCurrentReports() {
+  return createAsyncAction(types.FETCH_CURRENT_REPORTS, types.FETCH_CURRENT_REPORTS_COMPLETED, () => {
```
Comment (stasm) [inline www/src/app/actions/index.js]: nit (here and elsewhere): please use 80 characters as the max in one line

File: www/src/app/components/inbox/Inbox.js [hunk 11]
```diff
@@ -1,122 +1,126 @@
...
 }
+
+const mapStateToProps = (state) => ({
+  author: state.inbox.author,
+  nextReportDate: state.inbox.nextReportDate,
+  nextReportSlug: state.inbox.nextReportSlug
+});
+
+
+export default connect(mapStateToProps)(Inbox);
```
Comment (stasm) [inline www/src/app/components/inbox/Inbox.js]: I would prefer to `connect` the `InboxContainer` class rather than this stateless component.

File: www/src/app/components/inbox/UpdateList.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import * as actions from '../../actions';
+import { randElem } from '../../utils';
+
+class UpdateList extends Component {
```
Comment (stasm) [inline www/src/app/components/inbox/UpdateList.js]: As I mentioned on IRC, I think this deserves its own PR.  I'd prefer to change one big at a time and focus on introducing Redux in this PR.

File: www/src/app/components/inbox/Inbox.js [hunk 13]
```diff
@@ -1,122 +1,126 @@
...
-    </div>
-  );
+    );
+  }
 }
+
+const mapStateToProps = (state) => ({
+  author: state.inbox.author,
+  nextReportDate: state.inbox.nextReportDate,
+  nextReportSlug: state.inbox.nextReportSlug
```
Comment (stasm) [inline www/src/app/components/inbox/Inbox.js]: We also need `prevReportSlug` here to be able to put updates into the _Current goals_ scratchpad.

File: www/src/app/services/alfred-api-service.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import { API_URL } from '../../config';
+import { get, post } from '../utils';
+
+
+export function fetchCurrentUpdatesByAuthor(author) {
```
Comment (stasm) [inline www/src/app/services/alfred-api-service.js:5]: Factoring each call into a separate function is a great idea. +1

**Review by stasm — CHANGES_REQUESTED** [review]
I like the direction this is going, but I'd also like to keep this PR just about introducing Redux.  We can then follow up with more PRs and refactor the UpdateList and other parts too.

File: www/src/app/actions/index.js (hunk 9 above)
Comment (hkasemir) [reply www/src/app/actions/index.js]: sure! This is the convention we use at work, but request and receive are just as good! I'll update

Comment (hkasemir) [reply www/src/app/actions/index.js]: I think the only one that then looks out of place are patch and post actions, do you have thoughts  on that?

File: www/src/app/components/inbox/Inbox.js (hunk 13 above)
Comment (hkasemir) [reply www/src/app/components/inbox/Inbox.js]: ok, since I'm not setting that in a reducer yet - (I wanted to contain this to the initial `fetchCurrentReports` and `fetchUpdatesByAuthor` methods, neither of which seems to return the `prevReportSlug` - I think we can make sure to add that in a future PR.

Comment (stasm) [reply www/src/app/components/inbox/Inbox.js]: `fetchCurrentReports` always returns two reports: the previous one and the next one.

Comment (hkasemir) [reply www/src/app/components/inbox/Inbox.js]: ahh, k, you weren't setting it anywhere in `master` so I was confused. Will add.

Comment (stasm) [reply www/src/app/components/inbox/Inbox.js]: Ah, sorry, that's you get when I do reviews at midnight!  I meant `prevReportDate` not `prevReportSlug`.  Which was also missing from here when I did the review :)

File: www/src/app/actions/index.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+export function patchUpdate(update) {
+  return createAsyncAction(types.REQUEST_UPDATE_PATCH,
+                           types.RECEIVE_UPDATE_PATCH,
+                           () => {
+    return alf.patchUpdate(update).then(() => update);
+  });
+}
+
+export function postUpdate(update) {
```
Comment (stasm) [inline www/src/app/actions/index.js]: Here and elsewhere:  perhaps rename this to `createUpdate` and the corresponding actions to `_UPDATE_CREATE`?

File: www/src/app/reducers/inbox.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    const update = {
+      _id: Date.now(),
+      author: state.author,
+      reportDate: state.nextReportDate,
+      status: action.payload,
+      text: '',
+      resolved: false,
+      editable: true,
+      adding: true
+    };
```
Comment (stasm) [inline www/src/app/reducers/inbox.js]: nit: fix the indentation please.  Also, since you're binding a `const` here, I'd suggest wrapping the whole block in `{…}`:

```
case types.START_ADD: {
    const …
}
```

This will prevent any naming conflicts with other `const`s in this function.

File: www/src/app/actions/index.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+  return createAsyncAction(types.REQUEST_UPDATE_POST,
+                           types.RECEIVE_UPDATE_POST,
+                           () => {
+    return alf.postUpdate(update).then(() => update);
```
Comment (stasm) [inline www/src/app/actions/index.js]: Please use `makeUpdate` from `utils` here to avoid sending `editable` and `adding` via the `POST`.

File: www/src/app/views/InboxContainer.js [hunk 18]
```diff
@@ -30,114 +20,74 @@ export default class InboxContainer extends Component {
 
     this.handleResolve = this.handleResolve.bind(this);
     this.handleArchive = this.handleArchive.bind(this);
+
+    this.state = {
+      editText: '',
```
Comment (stasm) [inline www/src/app/views/InboxContainer.js:25]: Ah, cool, this is a good idea.  It allows us to not keep both the old text and the new text on the `update` while the editing input is active.

File: www/src/app/views/InboxContainer.js [hunk 19]
```diff
@@ -30,114 +20,74 @@ export default class InboxContainer extends Component {
...
-      })
-    ).catch(console.error);
+  componentWillMount() {
+    const { params, setAuthor } = this.props;
+    const { fetchCurrentReports, fetchUpdatesByAuthor } = this.props;
+
+    setAuthor(params.author);
+    fetchCurrentReports().then(reports => {
+      fetchUpdatesByAuthor(params.author, reports.nextReportSlug)
+    });
```
Comment (stasm) [inline www/src/app/views/InboxContainer.js]: This is again subjective, but for one-liners like this I prefer the syntax without braces:

```
fetchCurrentReports().then(
    reports => fetchUpdatesByAuthor(params.author, reports.nextReportSlug)
);
```

File: www/src/app/views/InboxContainer.js [hunk 20]
```diff
@@ -223,6 +159,28 @@ export default class InboxContainer extends Component {
...
         handleArchive={this.handleArchive}
       />
+      </section>
     );
   }
 }
+
+const mapStateToProps = (state) => ({
+  inbox: state.inbox,
+  updates: state.inbox.updates
```
Comment (stasm) [inline www/src/app/views/InboxContainer.js:169]: I like the separation between `inbox` and `updates` here, :+1: .

File: www/src/app/actions/index.js [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+  });
+}
+
+export function patchUpdate(update) {
+  return createAsyncAction(types.REQUEST_UPDATE_PATCH,
+                           types.RECEIVE_UPDATE_PATCH,
+                           () => {
+    return alf.patchUpdate(update).then(() => update);
+  });
+}
```
Comment (stasm) [inline www/src/app/actions/index.js:87]: Here and elsewhere: for one-liner callbacks I'd suggest using the implicit-return syntax:

```
export function patchUpdate(update) {
  return createAsyncAction(
    types.REQUEST_UPDATE_PATCH,
    types.RECEIVE_UPDATE_PATCH,
    () => alf.patchUpdate(update).then(() => update)
  );
}
```

**Review by stasm — CHANGES_REQUESTED** [review]
This looks great, thanks @hkasemir!

I have a few minor comments about naming and indentation, and one request to use `makeUpdate` to avoid passing too much data to the DB.

With these changes this looks ready to land :)

**Review by stasm — APPROVED** [review]
bravo

**Comment by stasm** [comment]
Merged and deployed!

celeb
